lib/archethic.ex

```diff
@@ -239,93 +239,53 @@ defmodule Archethic do
   end
 
-  @doc """
-  Retrieve a transaction chain based on an address from the closest nodes.
-  """
-  @spec get_transaction_chain(binary()) :: {:ok, list(Transaction.t())} | {:error, :network_issue}
-  def get_transaction_chain(address) when is_binary(address) do
-    nodes = Election.chain_storage_nodes(address, P2P.authorized_and_available_nodes())
-
-    # We directly check if the transaction exists and retrieve the genesis
-    # Otherwise we are requesting the genesis address remotly
-    genesis_address =
-      with ^address <- TransactionChain.get_genesis_address(address),
-           {:ok, genesis_address} <- TransactionChain.fetch_genesis_address_remotely(address) do
-        genesis_address
-      else
-        _ ->
-          address
-      end
-
-    %{transactions: local_chain, last_address: last_local_address} =
-      genesis_address
-      |> TransactionChain.scan_chain(address)
-      |> Enum.reduce_while(%{transactions: [], last_address: nil}, fn transaction, acc ->
-        # We stop at the desire the transaction
-        if acc.last_address == address do
-          {:halt, acc}
-        else
-          new_acc =
-            acc
-            |> Map.update!(:transactions, &(&1 ++ [transaction]))
-            # We log the last local address
-            |> Map.put(:last_address, transaction.address)
-
-          {:cont, new_acc}
-        end
-      end)
-
-    remote_chain =
-      if address != last_local_address do
-        address
-        |> TransactionChain.stream_remotely(nodes, last_local_address)
-        |> Stream.flat_map(& &1)
-        |> Enum.to_list()
-      else
-        []
-      end
-
-    {:ok, local_chain ++ remote_chain}
-  end
-
   @doc """
   Retrieve a transaction chain based on an address from the closest nodes
   by setting `paging_address as an offset address.
   """
-  @spec get_transaction_chain_by_paging_address(binary(), binary(), :asc | :desc) ::
+  @spec get_transaction_chain_by_paging_address(binary(), binary() | nil, :asc | :desc) ::
           {:ok, list(Transaction.t())} | {:error, :network_issue}
-  def get_transaction_chain_by_paging_address(address, paging_address, order)
-      when is_binary(address) and order in [:asc, :desc] do
-    nodes = Election.chain_storage_nodes(address, P2P.authorized_and_available_nodes())
-
-    try do
-      {local_chain, paging_address} =
-        with true <- paging_address != nil,
-             true <- TransactionChain.transaction_exists?(paging_address),
-             last_address when last_address != nil <-
-               TransactionChain.get_last_local_address(address),
-             true <- last_address != paging_address do
-          {TransactionChain.get_locally(last_address, paging_address, order), last_address}
+  def get_transaction_chain_by_paging_address(address, paging_address, :asc)
+      when is_binary(address) do
+    case get_last_transaction_address(address) do
+      {:ok, last_address} ->
+        with {local_chain, false, _} <-
```

Comment:
Am I the only one to have trouble with this kind of `with`? It takes me 2 minutes to read it everytime.

Reply:
It is like a case, but with multiple values. If we rewrite this with case statement it would be like this
```elixir
case TransactionChain.get(address, [], paging_state: paging_address, order: :asc) do
  {local_chain, true, _} ->
    {:ok, local_chain}

  {local_chain, _, _} ->
    case List.last(local_chain) do
      %Transaction{address: ^last_address} -> 
        {:ok, local_chain}

      _ ->
        last_address
            |> Election.chain_storage_nodes(P2P.authorized_and_available_nodes())
            |> TransactionChain.fetch_transaction_chain(last_address, paging_address, order: :asc)
    end
end
```